src/backend/distributed/planner/query_colocation_checker.c

```diff
@@ -269,20 +284,93 @@ WrapRteRelationIntoSubquery(RangeTblEntry *rteRelation)
 	newRangeTableRef->rtindex = 1;
 	subquery->jointree = makeFromExpr(list_make1(newRangeTableRef), NULL);
 
-	/* Need the whole row as a junk var */
-	Var *targetColumn = makeWholeRowVar(newRangeTableEntry, newRangeTableRef->rtindex, 0,
-										false);
+	Relation relation = relation_open(rteRelation->relid, AccessShareLock);
+	int numberOfAttributes = RelationGetNumberOfAttributes(relation);
+
+	bool shouldAssignDummyNullColumn = list_length(requiredAttributes) == 0;
+	bool assignedDummyNullColumn = false;
+	int attributeNumber = 1;
+	int resultNo = 1;
+	for (; attributeNumber <= numberOfAttributes; attributeNumber++)
+	{
+		Form_pg_attribute attributeTuple =
+			TupleDescAttr(relation->rd_att, attributeNumber - 1);
+		Var *targetColumn =
+			makeVar(newRangeTableRef->rtindex, attributeNumber, attributeTuple->atttypid,
+					attributeTuple->atttypmod, attributeTuple->attcollation, 0);
+		TargetEntry *targetEntry =
+			makeTargetEntry((Expr *) targetColumn, attributeNumber,
+							strdup(attributeTuple->attname.data), false);
+
+		if (shouldAssignDummyNullColumn && !assignedDummyNullColumn)
+		{
+			/*
+			 * in case there is no required column, we assign one dummy NULL target entry
+			 * to the subquery targetList so that it has at least one target.
+			 * (targetlist should have at least one element)
+			 */
+			subquery->targetList = lappend(subquery->targetList, targetEntry);
+			assignedDummyNullColumn = true;
+		}
 
-	/* create a dummy target entry */
-	TargetEntry *targetEntry = makeTargetEntry((Expr *) targetColumn, 1, "wholerow",
-											   true);
+		if (!list_member_int(requiredAttributes, attributeNumber))
+		{
+			/*
+			 * We add a null target entry because we don't have an easy
+			 * way of changing all the references to this column and
+			 * we don't want to break postgres query.
+			 */
+			targetEntry->expr =
+				(Expr *) makeNullConst(attributeTuple->atttypid,
+									   attributeTuple->atttypmod,
+									   attributeTuple->attcollation);
+			*allTargetList = lappend(*allTargetList, targetEntry);
+		}
+		else
+		{
+			TargetEntry *copyTargetEntry = copyObject(targetEntry);
+			*allTargetList = lappend(*allTargetList, copyTargetEntry);
+
+			/* In the subquery with only required attribute numbers, the result no
+			 * corresponds to the ordinal index of it in targetList.
+			 */
+			targetEntry->resno = resultNo++;
+			subquery->targetList = lappend(subquery->targetList, targetEntry);
+		}
+	}
+	MakeVarAttNosSequential(*allTargetList);
```

Comment:
why don't we set it to the right atnno when we generate the Var above?

Reply:
This method was quite unreadable and doing too many things, it wasn't possible to update the Var above but refactoring the method is easier than explaining why

https://github.com/citusdata/citus/pull/4358/commits/10dfee94c141cfdbcabbfc95a81bb3ef830dd837